File: services/api/app/routers/notifications.py
import asyncpg
from fastapi import APIRouter, Query, Response

from app.auth import CurrentUser
from app.clients import db
from app.queries import notifications as notifications_q
from app.schemas.notification import (
    NotificationItem,
    NotificationListResponse,
    NotificationsReadRequest,
)

router = APIRouter(prefix="/notifications")


def _item(row: asyncpg.Record) -> NotificationItem:
    return NotificationItem(
        notification_id=str(row["id"]),
        type=row["type"],
        payload=row["payload"],
        created_at=row["created_at"],
        read_at=row["read_at"],
    )


@router.get("", response_model=NotificationListResponse)
async def list_notifications(
    user: CurrentUser,
    limit: int = Query(default=50, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
) -> NotificationListResponse:
    rows, total, unread = await notifications_q.list_for_user(
        db.pool(), user.id, limit=limit, offset=offset
    )
    return NotificationListResponse(
        notifications=[_item(r) for r in rows], total=total, unread_count=unread
    )


@router.post("/read", status_code=204)
async def mark_read(body: NotificationsReadRequest, user: CurrentUser) -> Response:
    """Idempotent (3_api.md): already-read, foreign, and malformed ids no-op
    alike — none of them are the caller's unread notifications."""
    try:
        await notifications_q.mark_read(db.pool(), user.id, ids=body.ids, mark_all=body.all)
    except asyncpg.DataError:  # not a UUID
        pass
    return Response(status_code=204)

File: apps/web/src/lib/api/notifications.ts
import { apiFetch, apiSend } from "@/lib/api/client";

// Types mirror services/api/app/schemas/notification.py — keep in sync.

export type NotificationType = "review_request" | "subscription_match" | "upload_failed";

export type Notification = {
  notification_id: string;
  type: NotificationType;
  // Type-specific; always enough to build the link target.
  payload: Record<string, unknown>;
  created_at: string;
  read_at: string | null;
};

export type NotificationList = {
  notifications: Notification[];
  total: number;
  unread_count: number;
};

export async function listNotifications(limit = 50, offset = 0): Promise<NotificationList> {
  return apiFetch<NotificationList>(`/v1/notifications?limit=${limit}&offset=${offset}`);
}

export async function getUnreadCount(): Promise<number> {
  const { unread_count } = await listNotifications(1, 0);
  return unread_count;
}

export async function markNotificationsRead(
  body: { ids: string[] } | { all: true },
): Promise<void> {
  await apiSend("/v1/notifications/read", { method: "POST", body: JSON.stringify(body) });
}
